include/zephyr/dt-bindings/memory-attr/memory-attr.h

```diff
@@ -18,9 +18,10 @@
 #define DT_MEM_ATTR_SHIFT		(0)
 
 #define  DT_MEM_CACHEABLE		BIT(0)  /* cacheable */
-#define  DT_MEM_NON_VOLATILE		BIT(1)  /* non-volatile */
-#define  DT_MEM_OOO			BIT(2)  /* out-of-order */
-#define  DT_MEM_DMA			BIT(3)	/* DMA-able */
+#define  DT_MEM_NON_CACHEABLE		BIT(1)  /* non-cacheable */
```

Comment:
Previously we've discussed that we shouldn't simultaneously have `DT_MEM_CACHEABLE` and `DT_MEM_NON_CACHEABLE` bits. If there is no `DT_MEM_CACHEABLE` bit set that means that memory is uncachable.

I'm wondering about why do we bringing it back?

Reply:
Because you want to be able to specify that the memory region is actually non-cacheable and that you want to allocate memory from it.

Let's say that you don't have `DT_MEM_NON_CACHEABLE`, so you could have something like:

```
mem_cacheable: memory@10000000 {
       zephyr,memory-attr = <( DT_MEM_CACHEABLE | DT_MEM_ALLOCATABLE )>;
};

mem_non_cacheable: memory@20000000 {
       zephyr,memory-attr = <( DT_MEM_ALLOCATABLE )>;
};

mem_other: memory@30000000 {
       zephyr,memory-attr = <( DT_MEM_OOO | DT_MEM_DMA | DT_MEM_ALLOCATABLE )>;
};
```

How do you allocate memory from `mem_non_cacheable` that is non cacheable and at the same time being sure that you do not allocate from `mem_other`?
```c
mem_attr_heap_alloc( ???? , 0x100);
```

Reply:
> How do you allocate memory from mem_non_cacheable that is non cacheable and at the same time being sure that you do not allocate from mem_other?

And it's exactly my point :)

From what I see we are trying to use attributes not as attributes but as a labels.

In your example it would be logical if we allocate uncached memory from following nodes:

```
mem_non_cacheable: memory@20000000 {
       zephyr,memory-attr = <( DT_MEM_ALLOCATABLE )>;
};

mem_other: memory@30000000 {
       zephyr,memory-attr = <( DT_MEM_OOO | DT_MEM_DMA | DT_MEM_ALLOCATABLE )>;
};
```

But you don't want such behavior, so you introduce `DT_MEM_NON_CACHEABLE`.
It looks like a hack - as now memory can be `DT_MEM_CACHEABLE`, `DT_MEM_NON_CACHEABLE`, `not DT_MEM_CACHEABLE` and `not DT_MEM_NON_CACHEABLE`. Which is misleading.

But what's more important we'll bump into the same issue if multiple regions have `DT_MEM_NON_CACHEABLE`.

I.e., (using your example) we may want to have this dma memory to be uncached (i.e due to hw requirements). Or we may have tome another uncached region - i.e. for AMP membox. Or for million other usecases. And we mark them as `DT_MEM_NON_CACHEABLE` - as they are uncached (and we probably want that MPU / MMU set proper memory region attributes). So, we have following memory nodes.

```
mem_non_cacheable: memory@20000000 {
       zephyr,memory-attr = <( DT_MEM_NON_CACHEABLE | DT_MEM_ALLOCATABLE )>;
};

mem_other: memory@30000000 {
       zephyr,memory-attr = <( DT_MEM_NON_CACHEABLE | DT_MEM_OOO | DT_MEM_DMA | DT_MEM_ALLOCATABLE )>;
};
```

How do you allocate memory from mem_non_cacheable that is non cacheable and at the same time being sure that you do not allocate from mem_other? ;)

Reply:
Yes, I see the problem now and I see a couple of options here:
1. We introduce also the negative options like we did with `DT_MEM_CACHEABLE`, for example `DT_MEM_NON_DMA`, so you can allocate from `mem_non_cacheable` using `DT_MEM_NON_DMA |  DT_MEM_NON_CACHEABLE`
2. We introduce a specifier for the `DT_MEM_ALLOCATABLE`, for example `DT_MEM_ALLOCATABLE(CACHE) | DT_MEM_ALLOCATABLE(DMA) | DT_MEM_ALLOCATABLE(SPECIAL) | ...` even though that would be a bit redundant with the others `DT_MEM_*` attributes.

Reply:
Option 2 looks good to me.

Option 1 has multiple flaws. Probably most important is that lookup of required memory may be platform specific (i.e.: somewhere we need to pass `DT_MEM_NON_DMA | DT_MEM_NON_CACHEABLE`, somewhere is `DT_MEM_NON_CACHEABLE` enough and on some platform we may need even more complex mask). That would be issue if we want to do such allocation in generic code. Another issue is that we'll run out of available bits in `memory-attr` twice faster.

> Any other options you see?

Hmm. We can reuse another standard Zephyr solution and define link to required memory node in chosen, i.e.:

```
mem_non_cacheable: memory@20000000 {
       ...
};

chosen {
	zephyr,uncached-mem-pool = &mem_non_cacheable;
};
```

Before introducing memory attributes I'd do it this way.

Reply:
@evgeniy-paltsev your latest option doesn't seem scalable to me, otherwise we would add a different `chosen` option for each combination that could make sense? I'll work on option 1 and see how that looks.

Reply:
@evgeniy-paltsev I have pushed option 2 already for you to take a look 😉